src/pk_core.c

```diff
@@ -648,11 +648,37 @@ DEF(coreExit,
 // String functions.
 // -----------------
 
-// TODO: substring.
+DEF(coreStrSub,
+  "str_sub(str:string, pos:num, len:num) -> string\n"
+  "Returns a substring from a given string supplied. In addition, "
+  "the position and length of the substring are provided when this "
+  "function is called. For example: `str_sub(str, pos, len)`.") {
+
+  int64_t len, pos;
+
+  String* str = NULL;
+  if (!validateArgString(vm, 1, &str)) return;
+  if (!validateInteger(vm, ARG(2), &pos, "Argument 2")) return;
+  if (!validateInteger(vm, ARG(3), &len, "Argument 3")) return;
+
+  if (str->length < 1)
+    RET_ERR(newString(vm, "Expected a string length more than 0."));
+
+  if (str->length < pos || str->length < len)
+    RET_ERR(newString(vm, "Index out of range."));
```

Comment:
Sorry I've missed it in the last review, We should allow the position range to be `0 <= pos <= str->length`
when the `pos == 0` the user can still get a substring of length zero.

```c
if (pos < 0 || str->length < pos) RET_ERR(...);
```

The position should be added to the length check, if the length of the substring starting at the `pos` exceeded the string length it would be an error. (see below)

```c
if (str->length < pos + len) RET_ERR(...);
```

```
pos = 5
len = 7
                 v- str->length
"f o o b a r b a z"
           ^-----------^
           pos         pos+len
```


TODO:
I'm not expecting it from this pr but it's worth mentioning.
Make the length argument as optional to let the users allow get substring starting from `pos` to the reset of the string.

Reply:
@ThakeeNathees, thanks for mentioning. I'll send it in a follow up patch please.